src/tiledb/cloud/vcf/ingestion.py

```diff
@@ -0,0 +1,970 @@
+    :param vcf_attrs: VCF with all INFO/FORMAT fields to materialize, defaults to None
+    """
+
+    logger = get_logger()
+
+    graph = dag.DAG(
+        name="vcf-filter-uris",
+        namespace=namespace,
+        mode=dag.Mode.REALTIME,
+    )
+    submit = graph.submit_local if local_ingest else graph.submit
+
+    dataset_uri = submit(
+        create_dataset_udf,
+        dataset_uri,
+        config=config,
+        extra_attrs=extra_attrs,
+        vcf_attrs=vcf_attrs,
+        name="Create VCF dataset ",
+    )
+
+    if sample_list_uri:
+        sample_uris = submit(
+            read_uris_udf,
+            dataset_uri,
+            sample_list_uri,
+            config=config,
+            max_files=max_files,
+            name="Read VCF URIs ",
+        )
+
+    if search_uri:
+        sample_uris = submit(
+            find_uris_udf,
+            dataset_uri,
+            search_uri,
+            config=config,
+            pattern=pattern,
+            ignore=ignore,
+            max_files=max_files,
+            name="Find VCF URIs ",
+        )
+
+    sample_uris = submit(
+        filter_uris_udf,
+        dataset_uri,
+        sample_uris,
+        config=config,
+        name="Filter VCF URIs ",
+    )
+
+    run_dag(graph)
+
+    sample_uris = sample_uris.result()
+
+    if not sample_uris:
+        logger.info("All samples in the manifest have been ingested.")
+        return
+
+    logger.info("Found %d new URIs.", len(sample_uris))
+
+    graph = dag.DAG(
+        name="vcf-ingest-manifest",
+        namespace=namespace,
+        mode=dag.Mode.REALTIME,
+        max_workers=workers,
+    )
+    submit = graph.submit_local if local_ingest else graph.submit
+
+    # Adjust batch size to ensure at least 20 samples per worker
+    batch_size = min(batch_size, len(sample_uris) // workers)
+    batch_size = max(batch_size, 20)
+
+    num_partitions = ceil(len(sample_uris) / batch_size)
+    num_consolidates = ceil(num_partitions / workers)
+
+    consolidate = None
+    for i in range(num_partitions):
+        if i % workers == 0:
+            prev_consolidate = consolidate
+            consolidate = submit(
+                consolidate_dataset_udf,
+                dataset_uri,
+                config=config,
+                exclude=None,
+                include=[MANIFEST_ARRAY, LOG_ARRAY],
+                id=f"manifest-consol-{i//workers}",
+                name=f"Consolidate VCF Manifest {i//workers + 1}/{num_consolidates} ",
+            )
+
+        ingest = submit(
+            ingest_manifest_udf,
+            dataset_uri,
+            sample_uris[i * batch_size : (i + 1) * batch_size],
+            config=config,
+            id=f"manifest-ingest-{i}",
+            name=f"Ingest VCF Manifest {i+1}/{num_partitions} ",
+        )
+        if prev_consolidate:
+            ingest.depends_on(prev_consolidate)
+
+        if consolidate:
+            consolidate.depends_on(ingest)
```

Comment:
This is a little hard for me to follow. Is this designed to have N ingest steps, then a consolidate step, then N more ingest steps, then a consolidate step, etc.? It might be worth adding a comment explaining this.

Reply:
Thanks again @thetorpedodog. I addressed the latest comments and got rid of all the `shell=True` subprocesses.